### ISSUE TITLE
"Apps should not know about each other" principle

### ISSUE THREAD

**Issue by Mihaiii** [StarcounterApps/Content#106]
This is a general question, but I'll use Content app's case as an example.

Considering that one main value of Starcounter is that **one app should not know about another**, I have this case where TinyMCE's Image plugin (part of Content app)

![addimg](https://user-images.githubusercontent.com/8951553/30696189-1410afd2-9edb-11e7-9c3d-71de8c641949.png)

should work with Image app (if it's running):

![withimgapp](https://user-images.githubusercontent.com/8951553/30696217-2f3e8428-9edb-11e7-9866-fc3d4675cb42.png)

or without Image app:

![noimgapp](https://user-images.githubusercontent.com/8951553/30696268-5634f350-9edb-11e7-8f7f-b0b1f221c295.png)

Content app works with Image app like this: it accesses a URL that gets an empty page (inside Content app) and this page blends with the page from Images app. I need the empty page as a support for blending, but I don't need the empty page at all if no blending is done.

So basically in this case I would need to be aware of another app running, which is not in accordance with Starcounter's way of making apps.

My question is what would be the desired approach in this case?

//cc @LarsTjernstrom ,@warpech .

**Comment by LarsTjernstrom** [comment]
Hi @Mihaiii It's @warpech that's responsible for the blending that can help you.
Since I'm on a funeral tomorrow i can't attend a problem solving meeting until Monday efter lunch.

**Comment by Mihaiii** [comment]
I know that the simple answer is to include what's in the first image in a Content app page that blends with the image app. This is what was discussed with @warpech and it's tricky because tinymce generates the html and when someone writes a plugin for it, everything is in js. I'll have a look into how this can be done. It would also mean that I can't use all the logic that is already in TinyMCE's Image plugin.

**Comment by Mihaiii** [comment]
Suppose I move all that input tags and labels into a Content's page that would replace the empty page I'm now using (as I said in the previous message) and bind some data from Image app's main table to a `juicy-select`. 

This would respect the "apps should not know about each other" principle, but with a communication only through the database, the user experience would be very poor as there wouldn't really be a way to select an image via the Image app. Unless we add a new column in the Image app 's main table that would notify this selection. For what I need, the best user experience would imply an extra button with "Select" on the main page - that gets '/images' (after "Delete" and "Edit"). Of course, in this case, if the user runs only Image app (and no other app that uses `/images` - and I guess that at this moment Content app is the only one), he/she will not understand what that "Select" button is there for.

**Comment by Mihaiii** [comment]
In the last comment, I came with a proposed solution that would involve changes in Images app. Please let me know if I can make those changes.

**Comment by warpech** [comment]
Sorry, I think that the only way to go it to fork the Image plugin of TinyMCE to make it "Starcounter view" aware. 

Then you load what is on screenshot 1 to that plugin. Screenshot 2 is blended to it through the `Illustration` type with the following contexts:

```
Blender.MapUri<Illustration>("/content/illustration-picker/{?}", new string[]{"select", "whatis"});
```

This makes the apps not aware of each other, only aware of the data model.

You will need to create a new view in the Images app that serves as the illustration picker. `select` + `whatis` contexts say that this view will allow selecting image for the "WhatIs" part of the Illustration relation.

Illustration in Simplified is relation between ToWhat (in your case: WebContent) and WhatIs (in your case: the image).

Does this make sense? We can make a Hangout about this during the day.

**Comment by Mihaiii** [comment]
> You will need to create a new view in the Images app that serves as the illustration picker. select + whatis contexts say that this view will allow selecting image for the "WhatIs" part of the Illustration relation.

I don't quite understand what will happen if Images app is not running - what will be displayed in that case when the button in the first screenshot is pressed?

**Comment by Mihaiii** [comment]
After a talk we decided to use Search app to provided row content based on Image app.

Closing.
